vignettes/case-study-scaling.Rmd

```diff
@@ -0,0 +1,222 @@
+
+1. **Benchmark**: Pick a number *n*, how many users your application must concurrently support. Simulate *n* users interacting with your application and collect latency metrics. Use `shinyloadtest::record_session` to record a realistic interaction, and the `shinycannon` tool to simulate *n* users performing the recorded interaction.
+1. **Analyze**: Visualize and interpret the metrics. `shinyloadtest::load_runs` is used for this part. If it looks like *n* users can be supported, stop here. Otherwise, go to the next step.
+1. **Recommend**: Based on analysis, propose a way for the capacity of the application to be increased. Patterns in the metrics can help you identify what change in the application or its configuration would be most beneficial. It could be a code change, like [converting the app to async](https://rstudio.github.io/promises/articles/casestudy.html), or a configuration change, like increasing the number of R processes allocated to the application in RStudio Connect or Shiny Server Pro.
+1. **Optimize**: Implement one or more of the recommendations and benchmark again. Repeat the process until the application supports *n* users to your satisfaction.
+
+## The cranwhales demo app
+
+For the purposes of this case study, we'll pretend that [cranwhales](https://github.com/rstudio/cranwhales) is an app we wrote and that we're interested in ensuring it can support a minimum number of concurrent users. In a nutshell, cranwhales is a way to visualize anonymized download statistics from RStudio's `0-Cloud` CRAN mirror. For more on the app's features, see [the tour](https://rstudio.github.io/promises/articles/casestudy.html#a-tour-of-the-app).
+
+![](../docs/inst/img/cranwhales.png)
+
+cranwhales was written by [Joe Cheng](https://twitter.com/jcheng) to demonstrate Shiny's [async capabilities](https://blog.rstudio.com/2018/06/26/shiny-1-1-0/). We use it here because it's close to a typical Shiny app in size and scope, and because it's already given an optimization treatment in [Joe's async case study](https://rstudio.github.io/promises/articles/casestudy.html).
+
+Joe's case study focuses on the **optimize** step and the mechanics of improving performance with an emphasis on using async to do so. Here, we'll instead focus on the **benchmark** and **analyze** steps as enabled by `shinyloadtest` and `shinycannon`.
+
+## Deployment
+
+cranwhales is currently [deployed on shinyapps.io](https://gallery.shinyapps.io/cranwhales/), but we'll assume for this case study that you've deployed cranwhales to your own [RStudio Connect](https://www.rstudio.com/products/connect/) instance with default runtime/scheduler settings. `shinyloadtest` is capable of benchmarking and generating load against apps that require authentication (with some [caveats](load-testing-authenticated-apps.html)) but we'll assume your deployment of cranwhales is accessible without authentication.
```

Comment:
... I thought we couldn't benchmark against shinyapps.io?

Reply:
We can't, and I started that sentence with the intent of mentioning that, but forgot. Thanks for the catch.